### PR TITLE
Prefetching and Rendering on Netlify

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Changes to ISR, Render, and prefetching to support the Netlify environment

# Review By (Date)
- When you want

# Criticality
- Normal

# Review Tasks

## Setup tasks and/or behavior to test

1. Open up the deploy preview in Safari and Mobile
2. Navigate to the home page of the deploy preview
3. Click on any story link
4. Click on the logo to go back to the home page
5. Click the back button
6. Hover mouse over logo and/or scroll down the page

Bonus points, Check out storyblok editor as well.

### CODE REVIEW

**Comment by yvonnetangsu** [comment]
@sherakama I'm using Lighthouse and came across this message. Not sure if it would help with anything but just in case
Screenshot 2024-01-10 at 2 40 07 PM
https://developer.chrome.com/docs/lighthouse/performance/bf-cache/?utm_source=lighthouse&utm_medium=devtools

**Comment by sherakama** [comment]
Screenshot 2024-01-10 at 15 02 02


Thanks for that tip. I feel like I am narrowing in on the culprit. Might have to replace the react-player or adjust the settings from the external providers.

**Comment by sherakama** [comment]
Also found these

https://dev.to/alvesjessica/my-experience-deploying-a-nextjs-application-on-netlify-using-next-app-router-3k1b
https://github.com/vercel/next.js/discussions/50958

Seems others are having the same issue.

**Comment by yvonnetangsu** [comment]
> Also found these
> 
> https://dev.to/alvesjessica/my-experience-deploying-a-nextjs-application-on-netlify-using-next-app-router-3k1b [vercel/next.js#50958](https://github.com/vercel/next.js/discussions/50958)
> 
> Seems others are having the same issue.

Nice found! Arrrgggh frustrating 😞

**Comment by yvonnetangsu** [comment]
Looks like this latest round of disabling prefetch for every link does the trick!

**Comment by sherakama** [comment]
> Looks like this latest round of disabling prefetch for every link does the trick!

A few things have moved around with this version but you're now at a decision point.

1. This version where the back button issue is fixed and the home page prefetching is off. When you click on the logo you get the loading page for a bit before it returns you to normal. Other pages that can and are prefetched work much smoother and quicker.
2. The buggy back button issue but you get the logo prefetched and don't see the loading page.

That is the cross-roads that I am at right now. Which one would you prefer?

File: app/(storyblok)/[...slug]/not-found.tsx
Comment (sherakama) [inline app/(storyblok)/[...slug]/not-found.tsx:1]: 404 handler needs to be here as well now as all paths will now run the ISR handler but throw the 404 (which is an error at that boundary) if not found now.

File: app/(storyblok)/[...slug]/page.tsx [hunk 1]
```diff
@@ -16,11 +16,7 @@ type ParamsType = {
   slug: string[];
 };
 
-export const dynamicParams = false; // Don't generate pages at runtime.
-// Bug in Safari + Netlify + Next where back button doesn't function correctly and returns the user
-// back to the page they hit the back button on after scrolling or interacting with the page they went back to.
-// Setting a long revalidate time patches this until Next/Netlify fix the bug in future releases of their stuff.
-export const revalidate = 60 * 60 * 24 * 365;
+export const revalidate = false;
```
Comment (sherakama) [inline app/(storyblok)/[...slug]/page.tsx:19]: Cache forever!

File: app/(storyblok)/[...slug]/page.tsx [hunk 2]
```diff
@@ -107,11 +103,12 @@ async function getStoryData(params: { slug: string[] }) {
...
       }
       catch (e) {
-        throw error;
+        console.error('Error', error);
       }
     }
-    throw error;
   }
+
+  return { data: 404 };
```
Comment (sherakama) [inline app/(storyblok)/[...slug]/page.tsx:111]: Return a 404 if the page doesn't get data from Storyblok and handle the 404 error on the server component response.

File: app/(storyblok)/[...slug]/page.tsx [hunk 3]
```diff
@@ -121,18 +118,18 @@ export async function generateMetadata({ params }: { params: ParamsType }): Prom
   try {
     const { data } = await getStoryData(params);
     if (!data.story || !data.story.content) {
-      throw new Error(`No story data found for ${params.slug.join('/')}`);
+      notFound();
```
Comment (sherakama) [inline app/(storyblok)/[...slug]/page.tsx:121]: This throws an error but a special 404 error that gets picked up at the error boundary and returns the `not-found.tsx` file.

File: app/(storyblok)/error.tsx
Comment (sherakama) [inline app/(storyblok)/error.tsx:1]: Just made this look a little nicer.

File: app/global-error.tsx
Comment (sherakama) [inline app/global-error.tsx:1]: Brought this back as now the ISR and SSR functions technically run, however, hopefully only the ISR function will run and trigger 404s if something doesn't go right.

**Comment by sherakama** [comment]
Yarg. 

I swear I do test these. Cache cleared and updated to time-based again.

**Comment by yvonnetangsu** [comment]
> Yarg.
> 
> I swear I do test these. Cache cleared and updated to time-based again.

lol thank you for your patience!

**Review by yvonnetangsu — APPROVED** [review]
Tested on Safari desktop and mobile. Looks like it's all working! And editing in Storyblok still works (checked just in case) 🎉 Thank you for your patience!

**Pull request merged**